src/js/themes/media-theme-template.js

```diff
@@ -0,0 +1,117 @@
+import { defineCustomElement } from '../utils/defineCustomElement.js';
+import MediaTheme from './media-theme.js';
+import { InnerTemplatePart, TemplateInstance, AttrPart } from './template-parts.js';
+
+class MediaThemeTemplate extends MediaTheme {
+  constructor() {
+    super();
+  }
+
+  connectedCallback() {
+    this.template = this.getRootNode().querySelector(`#${this.getAttribute('template')}`);
+
+    this.templateInstance = new TemplateInstance(
+      this.template,
+      this.props,
+      mediaThemeProcessor
+    );
+
+    this.shadowRoot.append(this.templateInstance);
+
+    const observer = new MutationObserver(() => this.render());
+    observer.observe(this, { attributes: true });
+  }
+
+  get props() {
+    const props = {};
+    for (let attr of this.attributes) {
+      if (attr.value != null) {
+        props[camelCase(attr.name)] = attr.value === '' ? true : attr.value;
+      } else {
+        props[camelCase(attr.name)] = false;
+      }
+    }
+    return props;
+  }
+
+  render() {
+    this.templateInstance.update(this.props);
+  }
+}
+
+const camelCase = (name) => {
+  return name.replace(/[-_]([a-z])/g, ($0, $1) => $1.toUpperCase());
+};
+
+const mediaThemeProcessor = {
+  processCallback(instance, parts, state) {
+    if (!state) return;
+    for (const [expression, part] of parts) {
+      if (part instanceof InnerTemplatePart) {
+        console.log(part);
+        switch (part.directive) {
+          case 'set': {
+            // cache these so they return the same element across layouts
+            state[expression] = new TemplateInstance(part.template, state, mediaThemeProcessor);
+            break;
+          }
+          case 'layout': {
+            if (state.layout === part.expression) {
+              // todo: cache the `TemplateInstance` per layout
```

Comment:
What happens if these have to be recreated whenever needed?

Reply:
right now each time the layout is changed it recreates the HTML elements from the templates, it makes a clone basically. I imagined it could be cached based on the layout id which should be unique.